crates/rpc/rpc/src/eth/bundle.rs

```diff
@@ -0,0 +1,205 @@
+    pub fn new(eth_api: Eth, blocking_task_guard: BlockingTaskGuard) -> Self {
+        Self { inner: Arc::new(EthBundleInner { eth_api, blocking_task_guard }) }
+    }
+}
+
+impl<Eth> EthBundle<Eth>
+where
+    Eth: EthTransactions + 'static,
+{
+    /// Simulates a bundle of transactions at the top of a given block number with the state of
+    /// another (or the same) block. This can be used to simulate future blocks with the current
+    /// state, or it can be used to simulate a past block. The sender is responsible for signing the
+    /// transactions and using the correct nonce and ensuring validity
+    pub async fn call_bundle(&self, bundle: EthCallBundle) -> EthResult<EthCallBundleResponse> {
+        let EthCallBundle { txs, block_number, state_block_number, timestamp } = bundle;
+        if txs.is_empty() {
+            return Err(EthApiError::InvalidParams(
+                EthBundleError::EmptyBundleTransactions.to_string(),
+            ))
+        }
+        if block_number.to::<u64>() == 0 {
+            return Err(EthApiError::InvalidParams(
+                EthBundleError::BundleMissingBlockNumber.to_string(),
+            ))
+        }
+
+        let transactions =
+            txs.into_iter().map(recover_raw_transaction).collect::<Result<Vec<_>, _>>()?;
+
+        let (cfg, mut block_env, at) =
+            self.inner.eth_api.evm_env_at(state_block_number.into()).await?;
+
+        // need to adjust the timestamp for the next block
+        if let Some(timestamp) = timestamp {
+            block_env.timestamp = U256::from(timestamp);
+        } else {
+            block_env.timestamp += U256::from(12);
+        }
+
+        // use the block number of the request
+        block_env.number = U256::from(block_number);
+
+        self.inner
+            .eth_api
+            .spawn_with_state_at_block(at, move |state| {
+                let coinbase = block_env.coinbase;
+                let basefee = Some(block_env.basefee.to::<u64>());
+                let env = Env { cfg, block: block_env, tx: TxEnv::default() };
+                let db = CacheDB::new(StateProviderDatabase::new(state));
+
+                let initial_coinbase =
+                    DatabaseRef::basic(&db, coinbase)?.map(|acc| acc.balance).unwrap_or_default();
+                let mut coinbasebalance_before_tx = initial_coinbase;
+                let mut coinbasebalance_after_tx = initial_coinbase;
+                let mut total_gas_used = 0u64;
+                let mut total_gas_fess = U256::ZERO;
+                let mut hash_bytes = Vec::with_capacity(32 * transactions.len());
+
+                let mut evm = revm::EVM::with_env(env);
+                evm.database(db);
+
+                let mut results = Vec::with_capacity(transactions.len());
+                let mut transactions = transactions.into_iter().peekable();
+
+                while let Some(tx) = transactions.next() {
+                    let tx = tx.into_ecrecovered_transaction();
```

Comment:
perf: can do the recovery outside of the loop in parallel

Reply:
I don't think it is worth it considering that we expect a lot of concurrent requests.